multiprocess_functions.py

```diff
@@ -0,0 +1,166 @@
+  print(len(sorted_size_dict))
+  results = [None] * len(sorted_size_dict)
+
+  i = 0
+  for file, value in sorted_size_dict.items():
+    file_path = file
+    process_num = queue.pop(0)
+
+    if max_processes < numa_nodes:
+      if max_processes == 1:
+        if numa_nodes > 1:
+          mem = '0-{}'.format(numa_nodes-1)
+        else:
+          mem = '0'
+      else:
+        mem = '{}-{}'.format(str(process_num * (numa_nodes//max_processes)), str(((process_num + 1) * (numa_nodes//max_processes)) - 1))
+    else:
+      mem = str(process_num//(max_processes//numa_nodes))
+    
+    results[i] = pool.apply_async(bash_subprocess, args=(file_path, mem, core_list[process_num*cores_per_process: (process_num+1)*cores_per_process]), callback = update_queue)
+    i += 1
+    while len(queue) == 0 and i < len(sorted_size_dict):
+        time.sleep(0.05)
+  pool.close()
+  pool.join()
+
+  return error_files
+
+def create_process_list(files, MIN_MEM_PER_PROCESS, MIN_CORES_PER_PROCESS, LOAD_BALANCE_FACTOR):
+  total_cores = get_total_cores()
+  #numa_nodes
+  lscpu = subprocess.Popen(["lscpu"], stdout=subprocess.PIPE)
+  grep = subprocess.Popen(["grep", "NUMA node(s):"], stdin=lscpu.stdout, stdout=subprocess.PIPE)
+  awk = subprocess.Popen(["awk", "{print $3}"], stdin=grep.stdout, stdout=subprocess.PIPE)
+  #Get the output
+  numa_nodes = int(awk.communicate()[0])
+  cores_per_numa = total_cores//numa_nodes
+
+  #sockets
+  lscpu = subprocess.Popen(["lscpu"], stdout=subprocess.PIPE)
+  grep = subprocess.Popen(["grep", "Socket(s):"], stdin=lscpu.stdout, stdout=subprocess.PIPE)
+  awk = subprocess.Popen(["awk", "{print $2}"], stdin=grep.stdout, stdout=subprocess.PIPE)
+  #Get the output
+  sockets = int(awk.communicate()[0])
+
+  memory_per_numa_domain = check_available_memory()/numa_nodes
+  # Memory max
+  mm = memory_per_numa_domain // MIN_MEM_PER_PROCESS
+  # Core_max
+  cm = cores_per_numa // MIN_CORES_PER_PROCESS
+
+  pn = len(files) // numa_nodes
+  # Load_balance_max
+  lbm = max(pn//LOAD_BALANCE_FACTOR, 1)
+  print("Memory max {}, Core max {}, Load Balance max {}".format(mm, cm, lbm))
+  max_p = min(mm, cm, lbm)
+
+  # number which is 2^x and less than or equal to max
+  max_p = 2**int(math.log2(max_p))
+
+  max_processes_list = []
+  while max_p > 0:
+    max_processes_list.append(max_p*numa_nodes)
+    max_p = max_p // 2
+  if numa_nodes != sockets:
+    max_processes_list.append(sockets)
+  if numa_nodes != 1:
+    max_processes_list.append(1)
+  print("Max processes list: ", max_processes_list)
+
+  return max_processes_list
+
+def start_process_list(files, max_processes_list, bash_subprocess):
+  total_cores = get_total_cores()
+  for max_processes in max_processes_list:
+    os.environ["OMP_NUM_THREADS"] = str(total_cores//max_processes)
+    print("Number of OMP Threads = {}, for {} instances".format(os.environ.get('OMP_NUM_THREADS'), max_processes))
+    if len(files) >= max_processes:
```

Comment:
Let us also keep a TODO item to check what is the best policy in such scenarios: e.g. we have 5 proteins and 6 NUMA domains - should we run 5 proteins on 5 of the NUMA domains simultaneously or should we run 2 at a time on 2 sockets. This condition ensures that currently, we do the latter. 
We can have this code for now but look into this TODO at a later date.

Reply:
done